### DISCUSSION TITLE
Expose SObject Id in duplicate error object

### DISCUSSION THREAD

**Discussion by romainbenard** [jsforce/jsforce#1521] [Q&A]
Hi everyone 👋

I just tried the 3.2.2 version and I encountered an issue with my current `jsforce` usage.

To gain one request, my strategy was to `create` an Account and if it already exists, use the `error` object to retrieve its `id`.
To do that I took advantage of the `duplicateResult` key which is now missing in the `v3.2.1`
In this new version It means that I need to `find` first then a `create` or `update`

Is it an expected behavior in the v3? 

NB: I can't use `upsert` because I should not edit an existing account

Thanks!

**Comment by cristiand391** [comment]
> To do that I took advantage of the duplicateResult key which is now missing in the v3.2.1

was this working on a previous v3 version or are you using v1?

Also, can you share a code example of how you are creating the `Account` record? I see `DuplicateResult` is returned by the SOAP API but `connection.create`uses the REST API:
https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_objects_duplicateresult.htm

**Reply by romainbenard** [reply]
Thanks for your quick answer

I'm using v1 in prod and I wanted to update to v3 to fix security issues

I simplified the integration for the example but I litterally do something like 
```
const createAccount = async (
  conn: Connection,
  input: { [key: string]: string }
) => {
  const data = await conn.sobject('Account').create(input)

  return data
}
```

And to handle duplicate case I do something like this:
```
try {
   createAccount(conn, input)
}
catch(err) {  
   if (err.errorCode === 'DUPLICATES_DETECTED') {
     const accountId = err.duplicateResult.matchResults[0].matchRecords[0].record.Id
  
     return { accountId: err.duplicateResult.matchResults[0].matchRecords[0].record.Id }
   }
  ...
}
```

**Reply by romainbenard** [reply]
in v1 the error object is something like:
```
{
  errorCode: 'DUPLICATES_DETECTED'
  duplicateResult: {
    matchResults: {
      matchRecords: Record[]
    }[]
  }
}
```

in v3 i received
```
{
  errorCode: 'DUPLICATES_DETECTED'
  content: undefined 
}
```

**Reply by cristiand391** [reply]
Got it, v1 was setting all error props from the API as returned in the `Error` object:
https://github.com/jsforce/jsforce/blob/c04515846e91f84affa4eb87a7b2adb1f58bf04d/lib/http-api.js#L252

I got a fix for it but the error details will be in `error.content` (TODO: add this to the migration guide)

I'll get a PR up in a bit.

**Reply by cristiand391** [reply]
https://github.com/jsforce/jsforce/pull/1522

**Reply by romainbenard** [reply]
thanks a lot @cristiand391 🙏

Any idea about a new release date?

**Reply by cristiand391** [reply]
@romainbenard next week, we'll get this one before v3 GA (also planned for next week).

**Reply by cristiand391** [reply]
@romainbenard the fix for this is available in jsforce v3.2.3, see: https://github.com/jsforce/jsforce/discussions/1529

**Comment by romainbenard** [comment]
From what I read in the REST documentation, the duplicate should also be handled by the REST API if there is an error during creation.
https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_composite_sobjects_collections_create.htm?q=duplicate
